pkg/controller/metal3remediation/metal3remediation_controller.go

```diff
@@ -0,0 +1,470 @@
+			}
+
+			r.Log.Info("Remediation timed out and retry limit reached")
+
+			// When machine is still unhealthy after remediation, and it can be re-provisioned, delete the machine.
+			// Note: this differs to the upstream metal3 remediation, which sets a condition which is handled by CAPI
+			if ok, err := remediationMgr.CanReprovision(ctx); err != nil {
+				return ctrl.Result{}, errors.Wrapf(err, "Failed to check if machine can be reprovisoned")
+			} else if ok {
+				r.Log.Info("Deleting machine")
+				if err := remediationMgr.DeleteMachine(ctx); err != nil {
+					return ctrl.Result{}, errors.Wrapf(err, "Failed to delete machine")
+				}
+			} else {
+				r.Log.Info("Machine can't be re-provisioned, will not delete it")
+			}
+
+			// Remediation failed, so set unhealthy annotation on BMH
+			// This prevents BMH to be selected as a host.
+			err = remediationMgr.SetUnhealthyAnnotation(ctx)
+			if err != nil {
+				r.Log.Error(err, "error setting unhealthy annotation")
+				return ctrl.Result{}, errors.Wrapf(err, "error setting unhealthy annotation")
+			}
+
+			remediationMgr.SetRemediationPhase(infrav1.PhaseDeleting)
+			// no requeue, we are done
+			return ctrl.Result{}, nil
+
+		case infrav1.PhaseDeleting:
+			// nothing to do anymore
+			break
+
+		case infrav1.PhaseFailed:
+			// nothing to do anymore
+			break
+
+		default:
+			r.Log.Error(nil, "unknown phase!", "phase", remediationMgr.GetRemediationPhase())
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// remediateRebootStrategy executes the remediation using the reboot strategy.
+// Returns nil, nil when reconcile can continue.
+// Return a Result and optionally an error when reconcile should return.
+func (r *Metal3RemediationReconciler) remediateRebootStrategy(ctx context.Context,
+	remediationMgr baremetal.RemediationManagerInterface,
+	node *corev1.Node) (ctrl.Result, error) {
+	// add finalizer
+	if !remediationMgr.HasFinalizer() {
+		remediationMgr.SetFinalizer()
+		return ctrl.Result{RequeueAfter: 1 * time.Second}, nil
+	}
+
+	// power off if needed
+	if ok, err := remediationMgr.IsPowerOffRequested(ctx); err != nil {
+		r.Log.Error(err, "error getting poweroff annotation status")
+		return ctrl.Result{}, errors.Wrap(err, "error getting poweroff annotation status")
+	} else if !ok {
+		r.Log.Info("Powering off the host")
+		err = remediationMgr.SetPowerOffAnnotation(ctx)
+		if err != nil {
+			r.Log.Error(err, "error setting poweroff annotation")
+			return ctrl.Result{}, errors.Wrap(err, "error setting poweroff annotation")
+		}
+
+		// done for now, wait a bit before checking if we are powered off already
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	// wait until powered off
+	if on, err := remediationMgr.IsPoweredOn(ctx); err != nil {
+		r.Log.Error(err, "error getting power status")
+		return ctrl.Result{}, errors.Wrap(err, "error getting power status")
+	} else if on {
+		// wait a bit before checking again if we are powered off already
```

Comment:
Since you're using reboot, it's possible that you miss the moment when the host is powered off. Or am I misunderstanding something?

Reply:
The initial upstream implementation used a reboot indeed, but I had to refactor it to power on and power off explicitely when I added node deletion. We didn't change the name of the remediation strategy though, I guess this causes some confusion? 
See line 287 for power off, and line 170 for power on.